config/discovery.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	configFieldPrefix      = "AUTO_"
+	staticConfigsKey       = "static_configs"
+	staticConfigsFieldName = configFieldPrefix + staticConfigsKey
+)
+
+var (
+	configNames      = make(map[string]discovery.Config)
+	configFieldNames = make(map[reflect.Type]string)
+	configFields     []reflect.StructField
+	configTypes      sync.Map // map[reflect.Type]reflect.Type
```

Comment:
We don't use sync.Maps anywhere else in Prometheus, why here?

Reply:
Because I needed a read-optimized concurrency-safe map and there's one in the standard library.

Reply:
Performance isn't important here, plus sync.Map is only for certain very specific use cases. Use a normal map with a lock.

Reply:
Done.